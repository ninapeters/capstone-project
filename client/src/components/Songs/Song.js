import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import AudioButton from '../Buttons/AudioButton'
import Button from '../Buttons/Button'
import { SaveIcon, SkipIcon } from '../Icons'

Song.propTypes = {
  artists: PropTypes.array.isRequired,
  updateTrack: PropTypes.func.isRequired,
  toggleCurrentSongId: PropTypes.func.isRequired,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
  saveSong: PropTypes.func.isRequired,
  savedSongs: PropTypes.array,
}

export default function Song({
  artists,
  updateTrack,
  stopPlayingSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
  saveSong,
  savedSongs,
}) {
  return (
    <SongWrapper>
      {artists?.map(({ artist, songTitle, image, id }) => (
        <SongGrid key={id}>
          <Cover src={image} alt="" />
          <Content>
            <ArtistName>{artist}</ArtistName>
            <SongTitle>{songTitle}</SongTitle>
          </Content>
          <ButtonWrapper>
            <Button
              isRedIcon
              id={id}
              onClick={skipSong}
              data-testid="skip-button"
            >
              <SkipIcon />
            </Button>
            <AudioButton
              onClick={() => toggleCurrentSongId(id)}
              isSongPlaying={isSongPlaying && currentSongId === id}
              currentSongId={currentSongId}
            />
            <Button
              id={id}
              onClick={() => saveSong(id)}
              disabled={savedSongs?.some((song) => song.id === id)}
              data-testid="save-button"
            >
              <SaveIcon />
            </Button>
          </ButtonWrapper>
        </SongGrid>
      ))}
    </SongWrapper>
  )

  function skipSong() {
    stopPlayingSong()
    updateTrack()
  }
}

const SongWrapper = styled.section`
  height: 100vh;
`
const SongGrid = styled.div`
  display: grid;
  flex: 1 0 100%;
  grid-template-columns: 100%;
  grid-template-rows: 54vh auto;
  position: relative;
  row-gap: 46px;
  scroll-snap-align: start;
`
const Cover = styled.img`
  border-radius: 0 0 70px 70px;
  box-shadow: var(--shadow-dark);
  height: 54vh;
  object-fit: cover;
  position: relative;
  width: 100%;
`
const Content = styled.section`
  padding: 0 22px;
`
const ArtistName = styled.span`
  color: var(--primary-regular);
  display: block;
  font-size: 1.2em;
  font-weight: 700;
  margin-bottom: 8px;
  text-transform: uppercase;
`
const SongTitle = styled.span`
  color: var(--primary-regular);
  font-size: 1.6em;
  font-weight: 700;
  line-height: 1;
  text-transform: uppercase;
`
const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
  position: absolute;
  top: 36%;
  width: 100%;
`
