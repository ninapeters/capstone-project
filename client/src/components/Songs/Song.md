One Song:

```jsx
<Song
  artists={[
  artist: 'Unprocessed'
  songTitle: 'Real'
  id: 'a'
  image: 'https://i.scdn.co/image/ab67616d0000b273251b8f4c6183988f5060b6eb'
  ]}
  updateTrack={() => {}}
  toggleCurrentSongId={() => {}}
  isSongPlaying={false}
  currentSongId="a"
  saveSong={() => {}}
  savedSongs={[]}
/>
```
