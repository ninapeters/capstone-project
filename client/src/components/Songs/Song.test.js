import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import Song from './Song'

const testProps = {
  artists: [],
  updateTrack: jest.fn(),
  toggleCurrentSongId: jest.fn(),
  isSongPlaying: false,
  currentSongId: 'a',
  saveSong: jest.fn(),
  savedSongs: [],
}

describe('Song', () => {
  it('renders correctly', () => {
    const props = { ...testProps }
    const { container } = render(<Song {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls updateTrack correctly', () => {
    const updateTrackMock = jest.fn()
    const props = { ...testProps, updateTrack: updateTrackMock }
    const { getByTestId } = render(<Song {...props} />)
    const button = getByTestId('skip-button')
    user.click(button)
    expect(updateTrackMock).toHaveBeenCalled()
  })

  it('calls toggleCurrentSongId with the correct id', () => {
    const toggleCurrentSongIdMock = jest.fn()
    const props = { ...testProps, toggleCurrentSongId: toggleCurrentSongIdMock }
    const { getByTestId } = render(<Song {...props} />)
    const button = getByTestId('audio-button')
    user.click(button)
    expect(toggleCurrentSongIdMock).toHaveBeenCalledWith('a')
  })

  it('switches the appearance of play and pause button by changing the state isSongPlaying', () => {
    let props = { ...testProps, isSongPlaying: false }
    const { getByTitle, queryByTitle, rerender } = render(<Song {...props} />)
    expect(getByTitle('play')).toBeInTheDocument()

    props = { ...testProps, isSongPlaying: true }
    rerender(<Song {...props} />)
    expect(queryByTitle('play')).not.toBeInTheDocument()
    expect(getByTitle('pause')).toBeInTheDocument()
  })

  it('calls saveSong correctly', () => {
    const saveSongMock = jest.fn()
    const props = { ...testProps, saveSong: saveSongMock }
    const { getByTestId } = render(<Song {...props} />)
    const button = getByTestId('save-button')
    user.click(button)
    expect(saveSongMock).toHaveBeenCalled()
  })
})
