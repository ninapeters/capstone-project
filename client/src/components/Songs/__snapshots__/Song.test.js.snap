// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song renders correctly 1`] = `
.c0 {
  height: 100vh;
}

<section
  class="c0"
/>
`;
